contracts/cw1-whitelist/src/integration_tests.rs

```diff
@@ -0,0 +1,129 @@
+use crate::msg::{AdminListResponse, ExecuteMsg, InstantiateMsg, QueryMsg};
+use anyhow::{anyhow, Result};
+use assert_matches::assert_matches;
+use cosmwasm_std::testing::{mock_env, MockApi, MockStorage};
+use cosmwasm_std::{to_binary, Addr, CosmosMsg, Empty, QueryRequest, WasmMsg, WasmQuery};
+use cw1::Cw1Contract;
+use cw_multi_test::{App, AppResponse, BankKeeper, Contract, ContractWrapper, Executor};
+use derivative::Derivative;
+use lazy_static::lazy_static;
+
+lazy_static! {
+    static ref OWNER: String = "owner".to_string();
+}
+
+fn mock_app() -> App {
+    let env = mock_env();
+    let api = MockApi::default();
+    let bank = BankKeeper::new();
+    let storage = MockStorage::new();
+
+    App::new(api, env.block, bank, storage)
+}
+
+fn contract_cw1() -> Box<dyn Contract<Empty>> {
+    let contract = ContractWrapper::new(
+        crate::contract::execute,
+        crate::contract::instantiate,
+        crate::contract::query,
+    );
+    Box::new(contract)
+}
+
+#[derive(Derivative)]
+#[derivative(Debug)]
+pub struct Suite {
+    /// Application mock
+    #[derivative(Debug = "ignore")]
+    pub app: App,
+    /// cw1 whitelist contract address
+    pub whitelist: Cw1Contract,
+}
+
+impl Suite {
+    pub fn init(mutable: bool) -> Result<Suite> {
+        let mut app = mock_app();
+        let cw1_id = app.store_code(contract_cw1());
+
+        let whitelist = app
+            .instantiate_contract(
+                cw1_id,
+                Addr::unchecked(OWNER.clone()),
+                &InstantiateMsg {
+                    admins: vec![OWNER.clone()],
+                    mutable,
+                },
+                &[],
+                "Whitelist",
+                None,
+            )
+            .unwrap();
+
+        Ok(Suite {
+            app,
+            whitelist: Cw1Contract(whitelist),
+        })
+    }
+
+    pub fn freeze(&mut self, contract_addr: &Addr) -> Result<AppResponse> {
```

Comment:
I find this signatre missleading. This doesn't call freeze directly, this calls `execute`, and freeze is just message to execute to ensure that execution works correctly. Better would be to go with signature like:

```
fn execute<Msg: serde::DeserializeOwned>(&mut self, contract_addr: &Addr, msg: Msg) -> Result<AppResponse>
```

And then call it:

```
execute(address, ExecuteMsg::Freeze {})?;
```

Reply:
Done.